### PR TITLE
DOC: Deploy docs to gh-pages from CircleCI - with versioning

### LINKED ISSUE #46: Build documentation
Set up a documentation build. Possibly pushing to github pages directly from Circle. WDYT?

### PR DESCRIPTION
**Pull request by rwblair**
Closes #46.

### CODE REVIEW

**Comment by effigies** [comment]
@rwblair What's the status of this? Should we try to get this merged today for a 0.3.0 release, or push off for now?

**Comment by oesteban** [comment]
@rwblair, they finally migrated sphinx-versioning to sphinx-contrib

See https://github.com/sphinx-contrib/sphinxcontrib-versioning/pull/46#issuecomment-386867658

**Comment by effigies** [comment]
I think moving on to upgrading fMRIPrep's dependencies is probably more important than squeezing this in now. Sorry for the short turnaround question at lunch time.

**Comment by oesteban** [comment]
Outstanding work! Will review ASAP

**Comment by rwblair** [comment]
@oesteban take another look at the docs, I removed the title and modules list for the packages as per your request.

If this looks ok we can merge it, and then I'll open a new PR removing the test tags and having sphinx-versioning using master as its reference point instead of the ci_doc_deploy branch.

**Comment by rwblair** [comment]
Biggest error in the doc build is coming from /nipype/sphinxext/plot_workflow.py:

> Traceback (most recent call last):
>   File "/home/rwblair/miniconda2/envs/smriprep/lib/python3.7/site-packages/nipype/sphinxext/plot_workflow.py", line 485, in run_code
>     exec(code, ns)
>   File "<string>", line 3, in <module>
> ImportError: cannot import name 'BIDSLayout' from 'pybids' (unknown location)

**Comment by oesteban** [comment]
Perhaps adding the `autodocs_mock_imports` could help (https://github.com/poldracklab/fmriprep/blob/6f3e63a44329dd9560b3e0d01779b821c406c900/docs/conf.py#L50-L54)

Also, we could remove the import from here:

https://github.com/poldracklab/smriprep/blob/268647be2167a59ad8cb9606b50c93910ec0468f/smriprep/workflows/base.py#L58 

https://github.com/poldracklab/smriprep/blob/268647be2167a59ad8cb9606b50c93910ec0468f/smriprep/workflows/base.py#L196

and replace them with a `namedtuple` as in https://github.com/poldracklab/fmriprep/blob/master/fmriprep/workflows/base.py#L86

File: smriprep/workflows/base.py [hunk 1]
```diff
@@ -193,7 +193,7 @@ def init_single_subject_wf(
 
         from collections import OrderedDict
         from smriprep.workflows.base import init_single_subject_wf
-        from bids import BIDSLayout
+        BIDSLayout = namedtuple('BIDSLayout', ['root'])
```
Comment (oesteban) [inline smriprep/workflows/base.py]: I'm guessing you'll need to add above the import:
```Python
from collections import OrderedDict, namedtuple
```

**Comment by effigies** [comment]
Sorry... Had to change the title for my sanity. Feel free to change it to something more accurate.

File: .circleci/config.yml [hunk 2]
```diff
@@ -173,7 +173,8 @@ jobs:
       - run:
           name: Setup Python environment
           command: |
-            pyenv global 3.5.2
+            pyenv install 3.6.3
+            pyenv global 3.6.3
```
Comment (effigies) [inline .circleci/config.yml]: If we don't need to build docker images, then we can just do:

```diff
   test_pytest:
-    machine:
-      image: circleci/classic:201711-01
+    docker:
+      - image: circleci/python:latest
     working_directory: /home/circleci/src/smriprep
```

And skip the `pyenv`.

File: .circleci/config.yml [hunk 3]
```diff
@@ -173,7 +173,6 @@ jobs:
       - run:
           name: Setup Python environment
           command: |
-            pyenv global 3.5.2
```
Comment (effigies) [inline .circleci/config.yml]: Ah, damn. I'm sorry. Now all the `pip` installs need to be `--user`. I thought I was making things simpler...

Comment (rwblair) [reply .circleci/config.yml]: Thanks. Yeah moving away from the machine images will make things simpler in the long tun.

Comment (effigies) [reply .circleci/config.yml]: Will probably need it on L185, as well.

**Comment by rwblair** [comment]
Tests look fixed. Any objections to merging @oesteban? Once this is merged I'll open up another PR that removes the doc_test_* tags and removes the ci_doc_deploy branch from being built via sphinxcontrib-versioning.

**Comment by oesteban** [comment]
I'll have a last look later today :+1: . Thanks very much for figuring this out.

File: .circleci/config.yml (hunk 3 above)
Comment (oesteban) [reply .circleci/config.yml]: Alternatively, we can consider `venv` (https://github.com/nipreps/dmriprep/blob/ae9a46710ce72b7e335c2a76d6cbb22f4f54150a/.circleci/config.yml#L250-L265)

File: .circleci/config.yml [hunk 4]
```diff
@@ -158,8 +158,8 @@ jobs:
             - /tmp/ds005/derivatives/freesurfer
 
   test_pytest:
-    machine:
-      image: circleci/classic:201711-01
+    docker:
+      - image: circleci/python:latest
```
Comment (oesteban) [inline .circleci/config.yml]: Why not pin a version (e.g. python:3.7.4)?

File: .circleci/config.yml [hunk 5]
```diff
@@ -427,6 +426,53 @@ jobs:
...
+          command: apt update && apt -y install graphviz
+      - run:
+          name: Install deps
+          command: pip install -r requirements.txt && pip install -r docs/requirements.txt
+      - run:
+          name: Build only this commit
+          command: sphinx-build -W ./docs ./docs/_build/no_version_html
+      - run:
+          name: Generate Docs
+          command: sphinx-versioning -vv build -r ci_doc_deploy ./docs ./docs/_build/html/
```
Comment (oesteban) [inline .circleci/config.yml]: I don't see the Makefile used anywhere here? Shouldn't we have only one way of generating docs? (I'd like the `make`, btw)

File: docs/_templates/apidoc/package.rst_t.bak [hunk 6]
```diff
@@ -0,0 +1,52 @@
+{%- macro automodule(modname, options) -%}
```
Comment (oesteban) [inline docs/_templates/apidoc/package.rst_t.bak]: Should this .bak file be removed?

File: docs/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+Sphinx>=2.1.2
+git+https://github.com/rwblair/sphinxcontrib-versioning.git@39b40b0b84bf872fc398feff05344051bbce0f63#egg=sphinx-versioning
+sphinxcontrib-apidoc
+# workflows autodocs were failing without pytest
+pytest
```
Comment (oesteban) [inline docs/requirements.txt]: maybe exclude all test files and the `conftest.py` file? If that doesn't work, then add `pytest` to the mocked imports?

File: smriprep/workflows/anatomical.py [hunk 8]
```diff
@@ -47,7 +47,7 @@ def init_anat_preproc_wf(
       - Brain extraction and INU (bias field) correction.
       - Brain tissue segmentation.
       - Spatial normalization to standard spaces.
-      - Surface reconstruction with FreeSurfer_.
+      - Surface reconstruction with FreeSurfer.
```
Comment (oesteban) [inline smriprep/workflows/anatomical.py]: I think you need to include `links.rst` to be able to use the reference.

File: .circleci/config.yml (hunk 4 above)
Comment (rwblair) [reply .circleci/config.yml]: fmriprep build docs is using latest. Want me to pin it?

Comment (oesteban) [reply .circleci/config.yml]: I think pinning will preempt having sudden errored builds, it is a good idea IMHO

File: smriprep/workflows/anatomical.py (hunk 8 above)
Comment (rwblair) [reply smriprep/workflows/anatomical.py]: I haven't been able to figure out where the include should be called.

File: docs/requirements.txt (hunk 7 above)
Comment (rwblair) [reply docs/requirements.txt]: Removed it from the requirements and the build succeeded.

File: docs/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,4 @@
+Sphinx>=2.1.2
```
Comment (effigies) [inline docs/requirements.txt]: Can you add/update these in `setup.cfg` under the `doc` extra?

https://github.com/poldracklab/smriprep/blob/5b519b1e2446ec8b69848575d7bc67800474f035/setup.cfg#L48-L54

File: smriprep/workflows/anatomical.py (hunk 8 above)
Comment (effigies) [reply smriprep/workflows/anatomical.py]: Added a commit.

File: docs/requirements.txt (hunk 9 above)
Comment (effigies) [reply docs/requirements.txt]: A useful tool we could include here, to keep things synced: https://github.com/nipy/nibabel/blob/master/tools/update_requirements.py

We could update it to also write out to `docs/requirements.txt`.

**Comment by effigies** [comment]
Looks like you need `pybids ~= 0.9.4`

**Comment by oesteban** [comment]
READY????

**Pull request merged**